File: apps/api/app/services/llm_metrics.py
from __future__ import annotations

import math

from app.schemas import LLMEstimateRequest, LLMEstimateResponse

_BYTES_BY_PRECISION = {
    "fp32": 4,
    "fp16": 2,
    "bf16": 2,
    "int8": 1,
    "int4": 0.5,
}


def estimar_costos_llm(req: LLMEstimateRequest) -> LLMEstimateResponse:
    bytes_per_value = _BYTES_BY_PRECISION[req.precision]
    safe_heads = max(1, req.query_heads)
    safe_kv_heads = min(req.kv_heads, safe_heads)
    safe_swa_window = min(req.context_length, req.swa_window_size)

    kv_mha_gb = req.num_layers * req.context_length * 2 * req.dimension * bytes_per_value * req.batch_size / 1e9
    kv_gqa_gb = kv_mha_gb * (safe_kv_heads / safe_heads)
    kv_swa_gb = req.num_layers * safe_swa_window * 2 * req.dimension * bytes_per_value * req.batch_size / 1e9
    kv_mla_gb = req.num_layers * req.context_length * 2 * req.mla_rank * bytes_per_value * req.batch_size / 1e9
    relations = (req.context_length * req.context_length) / 1e6
    perplexity_proxy = 8 + math.log2(max(2, req.context_length / 1024)) * 0.35 + (0.15 if safe_kv_heads < safe_heads else 0)
    tokens_per_second_proxy = max(1, round(1800 / math.sqrt(req.context_length / 1024) * (safe_kv_heads / safe_heads + 0.35)))

    notes = [
        "Estimación didáctica, no benchmark real de GPU.",
        "GQA reduce KV cache compartiendo claves/valores entre query heads.",
        "SWA limita el contexto activo a una ventana local conceptual.",
        "MLA representa el KV cache mediante un rango latente comprimido.",
    ]
    if req.rope:
        notes.append("RoPE activado como codificación posicional conceptual.")

    return LLMEstimateResponse(
        kv_cache_mha_gb=round(kv_mha_gb, 6),
        kv_cache_gqa_gb=round(kv_gqa_gb, 6),
        kv_cache_swa_gb=round(kv_swa_gb, 6),
        kv_cache_mla_gb=round(kv_mla_gb, 6),
        gqa_vs_mha_ratio=round(kv_gqa_gb / max(kv_mha_gb, 1e-9), 6),
        swa_vs_mha_ratio=round(kv_swa_gb / max(kv_mha_gb, 1e-9), 6),
        mla_vs_gqa_ratio=round(kv_mla_gb / max(kv_gqa_gb, 1e-9), 6),
        long_context_relations_millions=round(relations, 3),
        perplexity_proxy=round(perplexity_proxy, 3),
        tokens_per_second_proxy=tokens_per_second_proxy,
        notes=notes,
    )


estimate_llm_costs = estimar_costos_llm

File: apps/web/src/core/llmMetrics.ts
// Estimaciones didácticas para el KV Cache Estimator.
// No sustituyen mediciones reales con hardware, perfiles CUDA o benchmarks.

import type { LlmMetrics, LlmMetricsConfig } from "../types";

export function calcularMetricasLlm(config: LlmMetricsConfig): LlmMetrics {
  const bytesPerFp16 = 2;
  const safeQueryHeads = Math.max(1, config.queryHeads);
  const safeKvHeads = Math.max(1, Math.min(config.kvHeads, safeQueryHeads));
  const safeWindow = Math.max(1, Math.min(config.windowSize, config.contextLength));

  // KV cache aproximado: capas * contexto * 2(K,V) * dimensión * bytes * batch.
  const kvCacheMhaGB =
    (config.layers * config.contextLength * 2 * config.dimension * bytesPerFp16 * config.batchSize) / 1e9;

  // GQA reduce KV cache compartiendo K/V entre grupos de heads de consulta.
  const kvCacheGqaGB = kvCacheMhaGB * (safeKvHeads / safeQueryHeads);

  // SWA conceptual: limita el contexto activo a una ventana local.
  const kvCacheSwaGB =
    (config.layers * safeWindow * 2 * config.dimension * bytesPerFp16 * config.batchSize) / 1e9;

  // MLA conceptual: cachea una representación latente comprimida.
  const kvCacheMlaGB =
    (config.layers * config.contextLength * 2 * config.mlaRank * bytesPerFp16 * config.batchSize) / 1e9;

  const longContextRelationsMillions = (config.contextLength * config.contextLength) / 1e6;

  // Métrica pedagógica: sube levemente con contexto largo y con KV muy comprimido.
  const approximatePerplexity =
    8 + Math.log2(Math.max(2, config.contextLength / 1024)) * 0.35 +
    (safeKvHeads < safeQueryHeads ? 0.15 : 0);

  // Métrica pedagógica: cae al crecer el contexto; mejora con más KV heads.
  const estimatedTokensPerSecond = Math.max(
    1,
    Math.round((1800 / Math.sqrt(config.contextLength / 1024)) * (safeKvHeads / safeQueryHeads + 0.35))
  );

  return {
    kvCacheMhaGB,
    kvCacheGqaGB,
    kvCacheSwaGB,
    kvCacheMlaGB,
    gqaVsMha: kvCacheGqaGB / Math.max(kvCacheMhaGB, 1e-9),
    swaVsMha: kvCacheSwaGB / Math.max(kvCacheMhaGB, 1e-9),
    mlaVsGqa: kvCacheMlaGB / Math.max(kvCacheGqaGB, 1e-9),
    longContextRelationsMillions,
    approximatePerplexity,
    estimatedTokensPerSecond
  };
}

export const computeLlmMetrics = calcularMetricasLlm;
